Updated ES search functions and authentication

vitaliyok opened the pull request:
Hi,

Here are proposed changes to ES search and authentications options:
1. Renamed and refactored the original search function
2. Added a function which would allow users to reuse scroll id if the the search fails
3. Added a function which sorting options, as recommended by ES
4. Added support for API key object (with encoded API key) as generated by ES
5. Added some more options for users to view index mapping and available indices in the search template

Commit:
Updated ES search functions and authentication

mart-r commented on cogstack.py:
```diff
@@ -1,158 +1,528 @@
+        with pd.option_context('display.max_colwidth', None):
+            return pd.DataFrame(index_aliases_coll, columns=['Index', 'Aliases'])
 
-    def get_docs_generator(self, index: List, query: Dict, es_gen_size: int=800, request_timeout: Optional[int] = 300):
```
There's a bit that's using this:

https://github.com/CogStack/working_with_cogstack/blob/main/medcat/3_run_model/run_model.py#L73

vitaliyok replied:
The method has been removed as but can be brought back. 
All of the new methods return a Pandas DataFrame but currently this function returns raw JSON which is converted to a list of tuples. It also uses ES _source object. In the new functions, I have excluded _source object from search results and only returning "fields", as recommended by Elastic. The problem is that all fields are arrays and values need to be joined in for the resulting DataFrame. 
I think, it would be possible to change the implementation to use new methods but create tuples from DataFrame instead of brining the old function back.

mart-r replied:
Yes, I think that's the right approach here.

I'm not saying that the methods I've tagged need to be reimplemented. All I'm trying to do is make sure that the code that uses them (i.e in the other notebooks and/or scripts) gets updated alongside the changes to `cogstack.py`. I.e if someone uses the scripts we provide (after this change), they don't error out because the they are out of sync from the loaded module(s).

mart-r commented on cogstack.py:
```diff
@@ -1,158 +1,528 @@
-            df = pd.DataFrame(temp_results)
-        return df
-    
-    def DataFrame(self, index: str, columns: Optional[List[str]] = None):
```
There's a few bits that are using this:
https://github.com/CogStack/working_with_cogstack/blob/main/medcat/3_run_model/run_model.py#L32
https://github.com/CogStack/working_with_cogstack/blob/main/medcat/2_train_model/1_unsupervised_training/unsupervised_medcattraining.py#L28

vitaliyok replied:
This is using eland DataFrame which is the same as Pandas DataFrame and can be re-implemented without eland.

mart-r commented on cogstack.py:
```diff
@@ -1,158 +1,528 @@
+            The username to use when connecting to Elasticsearch. If not provided, the user will be prompted to enter a username.
+        password : str, optional
+            The password to use when connecting to Elasticsearch. If not provided, the user will be prompted to enter a password.
+        apiKey : Dict, optional
```
Generaly, we want `snake_case` names for variables. So `api_key` would make more sense.

vitaliyok replied:
Ok. I have changed this.

mart-r commented on cogstack.py:
```diff
@@ -1,158 +1,528 @@
-                                                       api_key=api_key,
-                                                       verify_certs=False,
-                                                       timeout=timeout)
+                 apiKey: Dict = None):
```
Generaly, we want `snake_case` names for variables. So `api_key` would make more sense.

mart-r commented on cogstack.py:
```diff
@@ -1,158 +1,528 @@
-
-        if api_key and api:
-            self.elastic = elasticsearch.Elasticsearch(hosts=hosts,
-                                                       api_key=api_key,
```
There were a few bits that used this:
https://github.com/CogStack/working_with_cogstack/blob/main/medcat/2_train_model/1_unsupervised_training/unsupervised_medcattraining.py#L27
https://github.com/CogStack/working_with_cogstack/blob/main/medcat/3_run_model/run_model.py#L27

vitaliyok replied:
This can be changed to match the new parameters. I can implement this too.

mart-r reviewed (changes requested):
There's a few minor issues I see.
Mostly to do with the removed methods having been used in other parts of the project.

Though the gist of it seems to be fine. I haven't tested it out, but I'm sure it'll work if you've been using it at GSTT.

vitaliyok commented:
It looks like the current implementation is still using the old CogStack text field: "body_analysed". It should probably be renamed to "document_Content" or not use any specific field names in the code here.

mart-r commented:
> It looks like the current implementation is still using the old CogStack text field: "body_analysed". It should probably be renamed to "document_Content" or not use any specific field names in the code here.

The exepctation is _generally_ that the user provides the correct fields they're interested in. I'm pretty sure `body_analysed` serves as just an example.

With that said, if there's a more relevant, up to date example, we'd be better off using that indeed.

Commit:
Updated API key parameter name to match naming convention

Commit:
Ignore __pycache__ directory during commit

mart-r commented:
Could you take a look regards to v8 in KCH @vladd-bit

vladd-bit commented:
Feel free to merge this, we will eventually upgrade to ES9 anyway.

Commit:
Moved authentication from constructor to
separate methods for API and basic.

vitaliyok commented:
I have changed the logic for different types of authentication and removed the mixed approach (api key and basic) from constructor. Now there are two instance methods for corresponding auth type. This should make more obvious for users which type to use. This was discussed a few days ago.

mart-r commented on .gitignore:
```diff
@@ -11,3 +11,6 @@ data/cogstack_search_results/
 # Default environments
 venv
+
+# pythin cache folder
```
pythin haha

mart-r reviewed (changes requested):
I think this change makes a lot of sense!
Separating the authentication is good.

However, there's a few concerns I have:
1. There is no guarantee that the user of the class actually does the authentication
   - If they just init with hosts, they don't necessarily know that they need to call another method
   - They will just see everything fail once they start to use the other methods
   - I think it might be beneficial to have class methods that deal with the auth as well, something like:
       ```python
       @classmethod
       def with_basic_auth(cls, hosts: List[str], username: Optional[str] = None, password: Optional[str] = None) -> 'CogStack':
           cs = cls(hosts)
           cs.use_basic_auth(username, password)
           return cs
       @classmethod
       def with_api_key_auth(cls, hosts: List[str], api_key: Optional[Dict] = None) -> 'CogStack':
           cs = cls(hosts)
           cs.use_api_key_auth(api_key)
           return cs
       ```
2. There's still a few places within the repo where the old init is used. We can't just update the class and not the examples / scripts
    - https://github.com/vitaliyok/working_with_cogstack/blob/feature/access_cogstack/medcat/2_train_model/1_unsupervised_training/unsupervised_medcattraining.py#L27
    - https://github.com/vitaliyok/working_with_cogstack/blob/feature/access_cogstack/medcat/3_run_model/run_model.py#L27
    - https://github.com/vitaliyok/working_with_cogstack/blob/feature/access_cogstack/search/search_template.ipynb
3. You've removed some customisability such as the `timeout` option. And now it's hard coded as default 300. Perhaps we could (at the very least) define this as a class level constant so someone _could_ change it? I.e
    ```python
    class CogStack(object):
        ES_TIMEOUT = 300
        # and just refer to `self.ES_TIMEOUT` in `__connect`
    ```

Commit:
Added class methods to initialize ES connection and some code refactoring

vitaliyok commented:
I have added the class methods, etc.
About the references from other other scripts the old implementation, I have not included the old functionality into this version. As we discussed earlier, it might be better to review these scripts, in general, and make a decision whether to try to replicate the functionality, rewrite the whole thing or use old version for existing references and the new one to only to read data. Maybe, use as a version reference in the Notebook script.

Commit:
Added reversed cogstack module to prev version, added new module cogstack2 and added deprecation warning

Commit:
Added some more comments to explain a difference between search functions

vitaliyok commented:
For backward compatibility, there are now two version of search. The original one also has a deprecation warning.

mart-r commented:
@vitaliyok could you take a look at the typing issues raised by the CI?
I've copied it here:
```
cogstack2.py:150: error: Incompatible types in assignment (expression has type "Union[Any, str]", variable has type "Optional[dict[Any, Any]]")  [assignment]
cogstack2.py:152: error: Argument "api_key" to "__connect" of "CogStack" has incompatible type "Union[Any, str, tuple[Union[Any, str], dict[Any, Any]], None]"; expected "Union[str, tuple[str, str], None]"  [arg-type]
cogstack2.py:154: error: X | Y syntax for unions requires Python 3.10  [syntax]
cogstack2.py:200: error: X | Y syntax for unions requires Python 3.10  [syntax]
cogstack2.py:237: error: X | Y syntax for unions requires Python 3.10  [syntax]
cogstack2.py:263: error: Incompatible default for argument "include_fields" (default has type "None", argument has type "list[str]")  [assignment]
cogstack2.py:263: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
cogstack2.py:263: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
cogstack2.py:321: error: Unexpected keyword argument "request_timeout" for "count" of "Elasticsearch"  [call-arg]
cogstack2.py:337: error: X | Y syntax for unions requires Python 3.10  [syntax]
cogstack2.py:441: error: X | Y syntax for unions requires Python 3.10  [syntax]
cogstack2.py:445: error: X | Y syntax for unions requires Python 3.10  [syntax]
cogstack2.py:446: error: X | Y syntax for unions requires Python 3.10  [syntax]
/opt/hostedtoolcache/Python/3.9.23/x64/lib/python3.9/site-packages/elasticsearch/_sync/client/__init__.py:953: note: "count" of "Elasticsearch" defined here
cogstack2.py:403: error: Argument "fields" to "search" of "Elasticsearch" has incompatible type "Optional[list[str]]"; expected "Optional[Sequence[Mapping[str, Any]]]"  [arg-type]
cogstack2.py:511: error: Item "dict[Any, Any]" of "Union[dict[Any, Any], list[str]]" has no attribute "append"  [union-attr]
cogstack2.py:519: error: Argument "fields" to "search" of "Elasticsearch" has incompatible type "Optional[list[str]]"; expected "Optional[Sequence[Mapping[str, Any]]]"  [arg-type]
```

PS: The project is currently set to support python 3.9 through to 3.12. Which is in line with what the `medcat` library supports.

Commit:
Fixed compatibility issues with from Python3.9 and some refactoring

vitaliyok commented:
OK. I have tested with Python 3.9. Seems to be working but lets see what CI says this time.

mart-r commented:
The CI still has issues:
```
cogstack2.py:171: error: Argument "api_key" to "__connect" of "CogStack" has incompatible type "Any | str | tuple[Any | str | None, Any | str | None] | None"; expected "str | tuple[str, str] | None"  [arg-type]
cogstack2.py:191: error: Incompatible types in assignment (expression has type "Elasticsearch", variable has type "None")  [assignment]
cogstack2.py:196: error: "None" has no attribute "ping"  [attr-defined]
cogstack2.py:243: error: "None" has no attribute "indices"  [attr-defined]
cogstack2.py:286: error: "None" has no attribute "count"  [attr-defined]
cogstack2.py:350: error: Argument 1 to "scan" has incompatible type "None"; expected "Elasticsearch"  [arg-type]
cogstack2.py:358: error: "None" has no attribute "count"  [attr-defined]
cogstack2.py:365: error: Item "None" of "tqdm[dict[str, Any]] | None" has no attribute "bar_format"  [union-attr]
cogstack2.py:368: error: Item "None" of "tqdm[dict[str, Any]] | None" has no attribute "set_description"  [union-attr]
cogstack2.py:451: error: Incompatible types in assignment (expression has type "list[str] | None", variable has type "Sequence[Mapping[str, Any]]")  [assignment]
cogstack2.py:458: error: "None" has no attribute "search"  [attr-defined]
cogstack2.py:476: error: "None" has no attribute "scroll"  [attr-defined]
cogstack2.py:484: error: "None" has no attribute "clear_scroll"  [attr-defined]
cogstack2.py:491: error: "None" has no attribute "clear_scroll"  [attr-defined]
cogstack2.py:567: error: Incompatible types in assignment (expression has type "list[str] | None", variable has type "Sequence[Mapping[str, Any]]")  [assignment]
cogstack2.py:585: error: "None" has no attribute "search"  [attr-defined]
```

Commit:
Fixed type matching issues flagged during build

vitaliyok commented:
Fixed CI build issues but will see if it works during the next build.

mart-r commented:
Still reporting one issue:
```
cogstack2.py:356: error: Incompatible types in assignment (expression has type "None", variable has type "tqdm[Any]")  [assignment]
```

Commit:
Fixed yet another Incompatible type issue

vitaliyok commented:
Fixed the type issue raised by CI. 🤞

mart-r reviewed (approved):
Looks good to me! Thanks
